src/metabase/upload.clj

```diff
@@ -408,32 +407,43 @@
       auto-pk-indices
       (map (partial remove-indices auto-pk-indices)))))
 
+(defn- columns-with-auto-pk [columns]
+  (merge (ordered-map/ordered-map auto-pk-column-keyword ::auto-incrementing-int-pk) columns))
+
+(defn- create-auto-pk-column? [driver db]
+  (driver/database-supports? driver :upload-with-auto-pk db))
+
 (defn- load-from-csv!
   "Loads a table from a CSV file. If the table already exists, it will throw an error.
    Returns the file size, number of rows, and number of columns."
-  [driver db-id table-name ^File csv-file]
+  [driver db table-name ^File csv-file]
   (with-open [reader (bom/bom-reader csv-file)]
-    (let [[header & rows]         (without-auto-pk-columns (csv/read-csv reader))
-          {:keys [extant-columns generated-columns]} (detect-schema header (sample-rows rows))
-          cols->upload-type       (merge generated-columns extant-columns)
-          col-definitions         (column-definitions driver cols->upload-type)
-          csv-col-names           (keys extant-columns)
-          col-upload-types        (vals extant-columns)
+    (let [auto-pk?                (create-auto-pk-column? driver db)
+          [header & rows]         (cond-> (csv/read-csv reader)
+                                    auto-pk?
+                                    without-auto-pk-columns)
+          cols->upload-type       (detect-schema header (sample-rows rows))
+          col-definitions         (column-definitions driver (cond-> cols->upload-type
+                                                               auto-pk?
+                                                               columns-with-auto-pk))
+          csv-col-names           (keys cols->upload-type)
+          col-upload-types        (vals cols->upload-type)
           parsed-rows             (vec (parse-rows col-upload-types rows))]
       (driver/create-table! driver
-                            db-id
+                            (:id db)
                             table-name
                             col-definitions
-                            :primary-key [auto-pk-column-keyword])
+                            (when auto-pk?
+                              {:primary-key [auto-pk-column-keyword]}))
       (try
-        (driver/insert-into! driver db-id table-name csv-col-names parsed-rows)
+        (driver/insert-into! driver (:id db) table-name csv-col-names parsed-rows)
         {:num-rows          (count rows)
-         :num-columns       (count extant-columns)
-         :generated-columns (count generated-columns)
+         :num-columns       (count cols->upload-type)
+         :generated-columns (if auto-pk? 1 0)
```

Comment:
I'd prefer not to hardcode this assumption, given our naming `without-auto-pk-columnS`. Should be simple enough to just count how many columns get removed by that method.

Since we're unlikely to iterate on this in 49 and are forward-porting this separately, happy for this to get resolved in that latter PR.

Reply:
I just checked the snowplow schema and we don't even use this `generated-columns` data. I don't think it's useful either. I'll remove it in the 50 branch.

https://github.com/metabase/metabase/blob/936c185a874ff25a7ecc4aa69947e3902a058cf4/snowplow/iglu-client-embedded/schemas/com.metabase/csvupload/jsonschema/1-0-0#L11-L48

Reply:
Watch out, we actually [did](https://github.com/metabase/metabase/blob/master/snowplow/iglu-client-embedded/schemas/com.metabase/csvupload/jsonschema/1-0-3) add it to the snowplow schema.

It was actually added in 49 already, but there was a slip up where the schema was only updated in snowcatcloud and not committed to the repo.

Reply:
Agree that I don't see the point in publishing it though, but deprecating it would be a huge pain.

Reply:
Oh right. I'm okay hard-coding the assumption that there's one auto-pk column anyway. I can't imagine why we'd ever need more than one auto-pk column, and I doubt we're going to add any more generated columns. I think you misinterpreted why `without-auto-pk-columns` is plural - it's because there can be more than one column named `_mb_row_id` in the CSV, and they will all be ignored.

Reply:
We might add another type of generated column though. One example would be a batch number to track which upload / append the row came from  "¯\_(ツ)_/¯

Reply:
perhaps ¯\_(ツ)_/¯ this is all YAGNI anyway since we don't actually use this data for anything

we'd have to update this line too :D
https://github.com/metabase/metabase/pull/39821/files#diff-af3188166dbf14ec61d196f0e71de90a50f7e2c9de2c152d7cd2e20c6af4051bR695

Reply:
Busted 🤣